crates/starknet-devnet-server/src/subscribe.rs

```diff
@@ -0,0 +1,159 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use axum::extract::ws::{Message, WebSocket};
+use futures::stream::SplitSink;
+use futures::SinkExt;
+use serde::{self, Serialize};
+use starknet_types::rpc::block::BlockHeader;
+use tokio::sync::Mutex;
+
+use crate::api::json_rpc::error::ApiError;
+use crate::rpc_core::request::Id;
+
+pub type SocketId = u64;
+
+type SubscriptionId = i64;
+
+#[derive(Debug)]
+pub enum Subscription {
+    NewHeads,
+    TransactionStatus,
+    PendingTransactions,
+    Events,
+}
+
+#[derive(Debug, Clone, Serialize)]
+#[serde(untagged)]
+pub enum SubscriptionConfirmation {
+    NewHeadsConfirmation(SubscriptionId),
+    TransactionStatusConfirmation(SubscriptionId),
+    PendingTransactionsConfirmation(SubscriptionId),
+    EventsConfirmation(SubscriptionId),
+    UnsubscriptionConfirmation(bool),
+}
+
+#[derive(Debug, Clone, Serialize)]
+#[serde(untagged)]
+pub enum SubscriptionNotification {
+    NewHeadsNotification(BlockHeader),
+    // TransactionStatusNotification,
+    // PendingTransactionsNotification,
+    // EventsNotification,
+}
+
+impl SubscriptionNotification {
+    fn method_name(&self) -> &'static str {
+        match self {
+            SubscriptionNotification::NewHeadsNotification(_) => "starknet_subscriptionNewHeads",
+            // SubscriptionNotification::TransactionStatusNotification => {
+            //     "starknet_subscriptionTransactionStatus"
+            // }
+            // SubscriptionNotification::PendingTransactionsNotification => {
+            //     "starknet_subscriptionPendingTransactions"
+            // }
+            // SubscriptionNotification::EventsNotification => "starknet_subscriptionEvents",
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum SubscriptionResponse {
+    Confirmation { rpc_request_id: Id, result: SubscriptionConfirmation },
+    Notification { subscription_id: SubscriptionId, data: Box<SubscriptionNotification> },
+}
+
+impl SubscriptionResponse {
+    fn to_serialized_rpc_response(&self) -> serde_json::Value {
```

Comment:
why not already created structs RpcNotification and RpcResponse/JsonRpcResponse

Reply:
I noticed them while writing the code, just thought that the structures I wrote give better control over the situation.

Reply:
constructing the json object can be delegated to already created types. SubscriptionResponse will stay. Similar to JsonRpcResponse and DevnetResponse

Reply:
I don't like both approaches but can't figure out a workaround. Resolving